File: June2021/OutofBoundaryPaths.java
/*@author : Coding Decoded
 TC : O(n *m*maxxMoves)
 SC : O(n*m*maxMoves)
 question link : https://leetcode.com/problems/out-of-boundary-paths/
 */
class Solution {

  private Long[][][] memo = null;
  private int mod = 1000000007;

  public int findPaths(int m, int n, int maxMove, int startRow, int startColumn) {
    memo = new Long[m + 1][n + 1][maxMove + 1];
    return (int)(findPathsMemoHelper(m, n, maxMove, startRow, startColumn));
  }


  public long findPathsMemoHelper(int m, int n, int maxMove, int row, int col) {


    if (maxMove < 0) {
      return 0;
    }

    // boundary condition
    if (row == m || col == n || row < 0 || col <0) {
      return 1;
    }


    if (memo[row][col][maxMove] != null)
      return memo[row][col][maxMove];

    long total = 0;
    long left_maxMoveReduced = findPathsMemoHelper(m, n, maxMove - 1, row, col - 1);
    long right_maxMoveReduced = findPathsMemoHelper(m, n, maxMove - 1, row, col + 1);
    long up_maxMoveReduced = findPathsMemoHelper(m, n, maxMove - 1, row - 1, col);
    long down_maxMoveReduced = findPathsMemoHelper(m, n, maxMove - 1, row + 1, col);

    total =  (( left_maxMoveReduced + right_maxMoveReduced + down_maxMoveReduced + up_maxMoveReduced) % mod);
    memo[row][col][maxMove] = total;
    return total;
  }
}

/* @author : Darshika Jha
   question link : https://leetcode.com/problems/out-of-boundary-paths/
*/

class Solution {
  public int findPaths(int m, int n, int maxMove, int startRow, int startColumn) {
    HashMap<String,Long> hm = new HashMap<>();
    long res = findPathsHelper(m,n,maxMove,startRow,startColumn,hm);
    return (int) res;
  }

  public long findPathsHelper(int m, int n, int maxMove, int startRow, int startColumn,HashMap<String,Long> hm){
    if(maxMove < 0){
      return 0;
    }
    if(startRow < 0 || startColumn < 0 || startRow >= m || startColumn >= n){
      return 1;
    }
    if(hm.containsKey(startRow+";"+startColumn+";"+maxMove)){
      return hm.get(startRow+";"+startColumn+";"+maxMove);
    }

    long res = 0;

    res += (findPathsHelper(m,n,maxMove-1,startRow-1,startColumn,hm) %1000000007);
    res += (findPathsHelper(m,n,maxMove-1,startRow,startColumn+1,hm) %1000000007);
    res += (findPathsHelper(m,n,maxMove-1,startRow+1,startColumn,hm) %1000000007);
    res += (findPathsHelper(m,n,maxMove-1,startRow,startColumn-1,hm) %1000000007);
    hm.put(startRow+";"+startColumn+";"+maxMove,(res %1000000007));
    return (res %1000000007);
  }
}

File: June2021/C++/OutOfBoundaryPaths.cpp

// author - Shree lakshmi
// TC : o(maxMoves*m*n)
// SC: o(maxMoves*m*n)

class Solution {
public:
    int mod = 1000000007;
    int helper(int N, int m, int n,int moves, int x, int y, vector<vector<vector<int>>>& t){
        if(x<0||y<0||x>=m||y>=n)
            return 1;
        if(moves == N)
            return 0;
        if(t[moves][x][y]!=-1)
            return t[moves][x][y] % mod;
        int temp=0;
        temp = (temp+helper(N,m,n,moves+1,x,y-1,t))%mod; // moving left
        temp = (temp+helper(N,m,n,moves+1, x-1,y,t))%mod; //moving up
        temp = (temp+helper(N,m,n,moves+1, x+1, y,t))%mod; //moving below
        temp = (temp+helper(N,m,n,moves+1, x, y+1,t))%mod; // moving right
        t[moves][x][y] = temp;
        return t[moves][x][y]%mod;
    }
    int findPaths(int m, int n, int maxMove, int startRow, int startColumn) {
        //storing -1 as when we reach maxMoves we will return 0
        vector<vector<vector<int>>> t(maxMove+1,vector<vector<int>>(m+1,vector<int>(n+1,-1)));
        t[maxMove][m][n] = helper(maxMove,m,n,0,startRow,startColumn,t);
        return t[maxMove][m][n] % mod;
    }
};